examples/android/CHIPTool/app/src/main/java/com/google/chip/chiptool/clusterclient/OtaProviderClientFragment.kt

```diff
@@ -0,0 +1,354 @@
+import java.io.BufferedInputStream
+import java.io.File
+import java.io.FileInputStream
+import java.io.IOException
+import java.util.Optional
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+
+class OtaProviderClientFragment : Fragment() {
+  private val deviceController: ChipDeviceController
+    get() = ChipClient.getDeviceController(requireContext())
+
+  private lateinit var scope: CoroutineScope
+
+  private lateinit var addressUpdateFragment: AddressUpdateFragment
+
+  private var _binding: OtaProviderClientFragmentBinding? = null
+
+  private val otaProviderCallback = OtaProviderCallback()
+  private val binding
+    get() = _binding!!
+
+  override fun onCreateView(
+    inflater: LayoutInflater,
+    container: ViewGroup?,
+    savedInstanceState: Bundle?
+  ): View {
+    _binding = OtaProviderClientFragmentBinding.inflate(inflater, container, false)
+    scope = viewLifecycleOwner.lifecycleScope
+
+    deviceController.setCompletionListener(ChipControllerCallback())
+
+    addressUpdateFragment =
+      childFragmentManager.findFragmentById(R.id.addressUpdateFragment) as AddressUpdateFragment
+
+    binding.selectFirmwareFileBtn.setOnClickListener { selectFirmwareFileBtnClick() }
+    binding.announceOTAProviderBtn.setOnClickListener {
+      scope.launch { sendAnnounceOTAProviderBtnClick() }
+    }
+
+    deviceController.startOTAProvider(otaProviderCallback)
+    return binding.root
+  }
+
+  override fun onStart() {
+    super.onStart()
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R && !Environment.isExternalStorageManager()) {
+      Toast.makeText(
+          requireContext(),
+          "Require to Allow management of all files permission",
+          Toast.LENGTH_LONG
+        )
+        .show()
+      startActivity(
+        Intent(Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION).apply {
+          addCategory("android.intent.category.DEFAULT")
+          data = Uri.parse("package:${requireContext().packageName}")
+        }
+      )
+    } else if (!checkPermissionForReadExternalStorage()) {
+      requestPermissionForReadExternalStorage()
+    }
+  }
+
+  private fun selectFirmwareFileBtnClick() {
+    val intent = Intent(Intent.ACTION_GET_CONTENT)
+    val uri = Uri.parse(Environment.getExternalStorageDirectory().path + "/Download/")
+    intent.setDataAndType(uri, "*/*")
+    startActivityForResult(intent, 0)
+  }
+
+  override fun onActivityResult(requestCode: Int, resultCode: Int, intent: Intent?) {
+    super.onActivityResult(requestCode, resultCode, intent)
+    if (intent == null) {
+      return
+    }
+    val uri = intent.data
+    if (uri == null) {
+      Log.d(TAG, "onActivityResult : null")
+      return
+    }
+
+    val filename = getRealPathFromURI(intent.data!!)
+
+    requireActivity().runOnUiThread { binding.firmwareFileTv.text = filename }
+  }
+
+  private fun getRealPathFromURI(contentUri: Uri): String {
+    if (!contentUri.path!!.startsWith("/storage")) {
```

Comment:
For readability, I generally prefer to reduce the brace, and return early if possible and avoid !! if possible
if (contentUri.path!!.startsWith("/storage")
  return contentUri.path!!
val id = ....

Reply:
modified